src/Icicle/Pipeline.hs

```diff
@@ -0,0 +1,181 @@
+annotOfError e
+ = case e of
+    CompileErrorParse sp
+     -> Just
+      $ Parsec.errorPos sp
+    CompileErrorDesugar _
+     -> Nothing
+    CompileErrorCheck       e'
+     -> SC.annotOfError  e'
+    CompileErrorConvert     e'
+     -> STC.annotOfError e'
+    CompileErrorFlatten _
+     -> Nothing
+    CompileErrorProgram _
+     -> Nothing
+
+instance Pretty CompileError where
+ pretty e
+  = case e of
+     CompileErrorParse p
+      -> "Parse error:" <> line
+      <> indent 2 (text $ show p)
+     CompileErrorDesugar d
+      -> "Desugar error:" <> line
+      <> indent 2 (text $ show d)
+     CompileErrorCheck ce
+      -> "Check error:" <> line
+      <> indent 2 (pretty ce)
+     CompileErrorConvert ce
+      -> "Convert error:" <> line
+      <> indent 2 (pretty ce)
+     CompileErrorFlatten d
+      -> "Flatten error:" <> line
+      <> indent 2 (text $ show d)
+     CompileErrorProgram d
+      -> "Program error:" <> line
+      <> indent 2 (text $ show d)
+
+type Var        = SP.Variable
+type QueryTop'  = SQ.QueryTop Parsec.SourcePos Var
+type QueryTop'T = SQ.QueryTop (ST.Annot Parsec.SourcePos Var) Var
+type Program'   = Core.Program () Var
+
+sourceParse :: Text -> Text -> Either CompileError QueryTop'
+sourceParse base t
+ = mapLeft CompileErrorParse
+ $ SP.parseQueryTop (CommonBase.OutputName base) t
+
+sourceDesugar :: QueryTop' -> Either CompileError QueryTop'
+sourceDesugar q
+ = runIdentity
+ $ runEitherT
+ $ bimapEitherT CompileErrorDesugar snd
+ $ Fresh.runFreshT
+     (STD.desugarQT q)
+     (freshNamer "desugar")
+
+sourceReify :: D.Dictionary -> QueryTop' -> Either CompileError QueryTop'
+sourceReify d q
+ = do q'c    <- fst <$> sourceCheck d q
+      let q'r = reify q'c
+      sourceDesugar $ SQ.reannotQT ST.annAnnot q'r
+ where
+  reify q'
+     = snd
+     $ runIdentity
+     $ Fresh.runFreshT
+         (ST.transformQT STR.reifyPossibilityTransform q')
+         (freshNamer "reify")
+
+sourceCheck :: D.Dictionary -> QueryTop' -> Either CompileError (QueryTop'T, ST.Type Var)
+sourceCheck d q
+ = let d' = D.featureMapOfDictionary d
+   in  mapLeft CompileErrorCheck
+     $ snd
+     $ flip Fresh.runFresh (freshNamer "t")
+     $ runEitherT
+     $ SC.checkQT d' q
+
+
+sourceConvert :: D.Dictionary -> QueryTop'T -> Either CompileError Program'
+sourceConvert d q
+ = mapRight snd
+ $ mapLeft CompileErrorConvert conv
+ where
+  d'        = D.featureMapOfDictionary d
+  conv      = Fresh.runFreshT
+                (STC.convertQueryTop d' q)
+                (freshNamer "conv")
+
+
+sourceInline :: D.Dictionary -> QueryTop'T -> QueryTop'
+sourceInline d q
+ = SQ.reannotQT ST.annAnnot
+ $ inline q
+ where
+  funs      = M.map snd
+            $ D.dictionaryFunctions d
+  inline q' = snd
+            $ Fresh.runFresh
+                (STI.inlineQT funs q')
+                (freshNamer "inline")
+
+
+sourceParseConvert :: Text -> Text -> Either CompileError Program'
+sourceParseConvert base t
```

Comment:
don't know whether this is used anywhere. guess it should desugar and reify before converting